### PR TITLE
add functional performance functions snakemake

### PR DESCRIPTION
**Pull request by galengorski**
This is a cleaned up version of the functional performance calculations with a snakemake wrapper to run the functions. The three functions are:

- `create_cleaned_io_data` -- generate cleaned up data for functional performance calcs
- `calc_it_metrics_for_pairs` -- calculate TE, MI, correlation for pairs of sources and sinks during time periods specified
- `calc_functional_performance` -- using the TE calculated from `calc_it_metrics_for_pairs` calculate difference between observed and modeled and save to file

### CODE REVIEW

File: 03a_it_analysis/src/estuary_salinity_functional_performance_func.py [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+    fp_data = ml_sf_inputs.join(coawst_3_years, how = 'outer')
+    
+    fp_data.to_csv(os.path.join('03b_model\\out',run_id,'ML_COAWST_Results_Input_Cleaned.csv'))
+
+
+def calc_it_metrics_for_pairs(data_loc, sources, sinks, years, write_loc):
+    '''
+     Generate pairs of sources and sinks from the list for each year (or range of years) listed, 
+     then calculate the transfer entropy (TE), mutual information (MI), and correlation for those pairs at
+     time lags of 0 to 10 days (hard coded now). Critical values for TE and MI are calculated as well.
```
Comment (jds485) [inline 03a_it_analysis/src/estuary_salinity_functional_performance_func.py:65]: looks like 0-9 days based on output files.

**Review by jds485 — CHANGES_REQUESTED** [review]
This looks good and the code runs for me. I have some minor suggestions for comments, and one suggestions to generalize to N replicates instead of hard coding for 5. Up to you if that makes sense to address.

**Comment by galengorski** [comment]
Thanks for the review @jds485 ! I added a function to generalize the number of replicates and I cleaned up some comments as you suggested. If you're ok with it, I'll go ahead and merge it.

File: 03a_it_analysis/src/estuary_salinity_functional_performance_func.py [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+
+    '''
+    #reps = ['00','01','02','03','04']
+    reps = find_reps(run_id)
+    ml_sf = pd.read_csv(os.path.join('03b_model/out',run_id,reps[0],'ModelResults.csv'), parse_dates = True, index_col = 'Unnamed: 0')
+    for i,rep in enumerate(reps):
+        temp = pd.read_csv(os.path.join('03b_model/out',run_id,rep,'ModelResults.csv'), parse_dates = True, index_col = 'Unnamed: 0')
+        ml_sf['saltfront_pred_'+rep] = temp['saltfront_pred'].copy()
+
+    ml_sf_mean = ml_sf[['saltfront_pred','saltfront_pred_01','saltfront_pred_02','saltfront_pred_03','saltfront_pred_04']].mean(axis = 1)
```
Comment (jds485) [inline 03a_it_analysis/src/estuary_salinity_functional_performance_func.py:61]: This also needs adjusting for number of reps, right?

Comment (galengorski) [reply 03a_it_analysis/src/estuary_salinity_functional_performance_func.py]: Yep! Great point, tested it with 5 replicates and it worked. Ok should be ready now

Comment (jds485) [reply 03a_it_analysis/src/estuary_salinity_functional_performance_func.py]: okay, you can merge

**Pull request merged**